code/modules/pai/software.dm

```diff
@@ -0,0 +1,247 @@
+/mob/living/silicon/pai/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PaiInterface", name)
+		ui.open()
+		ui.set_autoupdate(FALSE)
+
+/mob/living/silicon/pai/ui_data(mob/user)
+	var/list/data = list()
+	data["door_jack"] = hacking_cable
+	data["image"] = card.emotion_icon
+	data["installed"] = installed_software
+	data["ram"] = ram
+	return data
+
+/mob/living/silicon/pai/ui_static_data(mob/user)
+	var/list/data = list()
+	data["available"] = available_software
+	data["directives"] = laws.supplied
+	data["emagged"] = emagged
+	data["languages"] = languages_granted
+	data["master_name"] = master_name
+	data["master_dna"] = master_dna
+	return data
+
+/mob/living/silicon/pai/ui_act(action, list/params, datum/tgui/ui)
+	. = ..()
+	if(.)
+		return
+	if(available_software[action] && !installed_software.Find(action))
+		to_chat(usr, span_warning("You do not have this software installed."))
+		CRASH("[usr] attempted to activate software they hadn't installed: [action]")
```

Comment:
This shouldn't cause a runtime. Players can pass garbage data in, and this should be expected